lib/workers.js

```diff
@@ -0,0 +1,309 @@
+   */
+  createGroupsOfTests(numberOfWorkers) {
+    const files = this.codecept.testFiles;
+    const mocha = Container.mocha();
+    mocha.files = files;
+    mocha.loadFiles();
+    const groups = populateGroups(numberOfWorkers);
+    let groupCounter = 0;
+
+    mocha.suite.eachTest((test) => {
+      const i = groupCounter % groups.length;
+      if (test) {
+        const { id } = test;
+        groups[i].push(id);
+        groupCounter++;
+      }
+    });
+    return groups;
+  }
+
+  /**
+   * @param {Number} numberOfWorkers
+   */
+  createGroupsOfSuites(numberOfWorkers) {
+    const files = this.codecept.testFiles;
+    const mocha = Container.mocha();
+    mocha.files = files;
+    mocha.loadFiles();
+    const groups = populateGroups(numberOfWorkers);
+
+    mocha.suite.suites.forEach((suite) => {
+      const i = indexOfSmallestElement(groups);
+      suite.tests.forEach((test) => {
+        if (test) {
+          const { id } = test;
+          groups[i].push(id);
+        }
+      });
+    });
+    return groups;
+  }
+
+  /**
+   * @param {Object} config
+   */
+  overrideConfig(config) {
+    for (const worker of this.workers) {
+      worker.addConfig(config);
+    }
+  }
+
+  run() {
+    return new Promise((res, rej) => {
+      runHook(this.codecept.config.bootstrapAll, () => {
+        for (const worker of this.workers) {
+          const workerThread = createWorker(worker);
+          this._listenWorkerEvents(workerThread);
+        }
+        res(true);
+      }, 'bootstrapAll');
+    });
+  }
+
+  /**
+   * @returns {Array<WorkerObject>}
+   */
+  getWorkers() {
+    return this.workers;
+  }
+
+  /**
+   * Share data to Workers
+   *
+   * @param {Object} data
+   */
+  share(data) {
+    Container.share(data);
+  }
+
+  /**
+   * @returns {Boolean}
+   */
+  isFailed() {
+    return (this.stats.failures || this.errors.length) > 0;
+  }
+
+  _listenWorkerEvents(worker) {
+    worker.on('message', (message) => {
+      output.process(message.workerIndex);
+      switch (message.event) {
+        case event.test.failed: this.emit(event.test.failed, repackTest(message.data)); break;
+        case event.test.passed: this.emit(event.test.passed, repackTest(message.data)); break;
+        case event.test.finished: this.emit(event.test.finished, repackTest(message.data)); break;
+        case event.all.after: this._appendStats(message.data); break;
+      }
+    });
+
+    worker.on('error', (err) => {
+      this.errors.push(err);
+    });
+
+    worker.on('exit', () => {
+      this.closedWorkers += 1;
+      if (this.closedWorkers === this.numberOfWorkers) {
+        runHook(this.codecept.config.teardownAll, () => {
+          this.emit(event.all.result, !this.isFailed());
+        });
+      }
+    });
+  }
+
+  _appendStats(newStats) {
+    this.stats.passes += newStats.passes;
+    this.stats.failures += newStats.failures;
+    this.stats.tests += newStats.tests;
+  }
+}
+
+module.exports = Workers;
```

Comment:
add here:

```js
module.exports.WorkerStorage = WorkerStorage;
```
so we could access WorkerStorage singleton

Reply:
Added as a separate file to avoid circular dependency.